src/moscot/problems/base/_utils.py

```diff
@@ -228,3 +236,242 @@ def _filter_kwargs(*funcs: Callable[..., Any], **kwargs: Any) -> Dict[str, Any]:
+    seed
+        Random seed if ``method = 'perm_test'``.
+
+    %(parallel_kwargs)s
+
+    Returns
+    -------
+    Dataframe of shape ``(n_genes, 5)`` containing the following columns, one for each lineage:
+        - ``corr`` - correlation between the count data and push/pull distributions.
+        - ``pval`` - calculated p-values for double-sided test.
+        - ``qval`` - corrected p-values using Benjamini-Hochberg method at level `0.05`.
+        - ``ci_low`` - lower bound of the ``confidence_level`` correlation confidence interval.
+        - ``ci_high`` - upper bound of the ``confidence_level`` correlation confidence interval.
+    """
+
+    corr, pvals, ci_low, ci_high = _correlation_test_helper(
+        X.T,
+        Y.values,
+        method=method,
+        n_perms=n_perms,
+        seed=seed,
+        confidence_level=confidence_level,
+        **kwargs,
+    )
+    invalid = (corr < -1) | (corr > 1)
+    if np.any(invalid):
+        logger.warning(
+            f"Found `{np.sum(invalid)}` correlation(s) that are not in `[0, 1]`. "
+            f"This usually happens when gene expression is constant across all cells. "
+            f"Setting to `NaN`"
+        )
+        corr[invalid] = np.nan
+        pvals[invalid] = np.nan
+        ci_low[invalid] = np.nan
+        ci_high[invalid] = np.nan
+
+    res = pd.DataFrame(corr, index=feature_names, columns=[f"{c}_corr" for c in Y.columns])
+    for idx, c in enumerate(Y.columns):
+        p = pvals[:, idx]
+        valid_mask = ~np.isnan(p)
+
+        res[f"{c}_pval"] = p
+        res[f"{c}_qval"] = np.nan
+        if np.any(valid_mask):
+            res.loc[feature_names[valid_mask], f"{c}_qval"] = multipletests(p[valid_mask], alpha=0.05, method="fdr_bh")[
+                1
+            ]
+        res[f"{c}_ci_low"] = ci_low[:, idx]
+        res[f"{c}_ci_high"] = ci_high[:, idx]
+
+    # fmt: off
+    res = res[[f"{c}_{stat}" for c in Y.columns for stat in ("corr", "pval", "qval", "ci_low", "ci_high")]]
+    return res.sort_values(by=[f"{c}_corr" for c in Y.columns], ascending=False)
+    # fmt: on
+
+
+def _correlation_test_helper(
+    X: ArrayLike,
+    Y: ArrayLike,
+    method: CorrTestMethod = CorrTestMethod.FISCHER,
+    n_perms: Optional[int] = None,
+    seed: Optional[int] = None,
+    confidence_level: float = 0.95,
+    **kwargs,
+) -> Tuple[ArrayLike, ArrayLike, ArrayLike, ArrayLike]:
+    """
+    Compute the correlation between rows in matrix ``X`` columns of matrix ``Y``.
+
+    Parameters
+    ----------
+    X
+        Array or matrix of `(M, N)` elements.
+    Y
+        Array of `(N, K)` elements.
+    method
+        Method for p-value calculation.
+    n_perms
+        Number of permutations if ``method='perm_test'``.
+    seed
+        Random seed if ``method = 'perm_test'``.
+    confidence_level
+        Confidence level for the confidence interval calculation. Must be in `[0, 1]`.
+
+    %(parallel_kwargs)s
```

Comment:
Again not interpolated, though not as much of a problem since this is a private function.